cmd/erasure-multipart.go

```diff
@@ -587,21 +577,32 @@ func (er erasureObjects) PutObjectPart(ctx context.Context, bucket, object, uplo
 	rctx := rlkctx.Context()
 	defer uploadIDRLock.RUnlock(rlkctx.Cancel)
 
-	data := r.Reader
-	// Validate input data size and it can never be less than zero.
-	if data.Size() < -1 {
-		logger.LogIf(rctx, errInvalidArgument, logger.Application)
-		return pi, toObjectErr(errInvalidArgument)
-	}
-
 	uploadIDPath := er.getUploadIDDir(bucket, object, uploadID)
-
 	// Validates if upload ID exists.
 	fi, _, err := er.checkUploadIDExists(rctx, bucket, object, uploadID, true)
 	if err != nil {
 		return pi, toObjectErr(err, bucket, object, uploadID)
 	}
 
+	// Write lock for this part ID, only hold it if we are planning to read from the
+	// streamto avoid any concurrent updates.
+	//
+	// Must be held throughout this call.
+	partIDLock := er.NewNSLock(bucket, pathJoin(object, uploadID, strconv.Itoa(partID)))
+	plkctx, err := partIDLock.GetLock(ctx, globalOperationTimeout)
```

Comment:
```suggestion
	plkctx, err := partIDLock.GetLock(rctx, globalOperationTimeout)
```
This will cancel this, if the above gets unlocked.

Reply:
do we want it to be dependent like this? @klauspost - kept them independent as before.

Reply:
Ping @vadmeste (lock expert). If we have two locks in the same handler, should we chain them?

Reply:
> Ping @vadmeste (lock expert). If we have two locks in the same handler, should we chain them?

Currently chaining does not have an effect. You can keep that way @harshavardhana and I will send another PR to support chaining context for locks.